exec_test.go

```diff
@@ -331,45 +276,126 @@ func TestCommandStartWait(t *testing.T) {
 				return
 			}
 
+			if stdout != nil {
+				assert.Equal(t, tc.wantStdout, stdout.String(), "Mismatch in piped stdout")
+			}
+			if stderr != nil {
+				assert.Equal(t, tc.wantStderr, stderr.String(), "Mismatch in piped stderr")
+			}
+
 			err = cmd.Wait()
 			require.Error(t, err, "Unexpected success calling (*Cmd).Wait twice")
 		})
 	}
 }
 
+// bufferPipeOutput buffers the output stream of a command with an intermediate pipe.
+func bufferPipeOutput(t *testing.T, cmd *wsl.Cmd, pipeName string) *bytes.Buffer {
+	t.Helper()
+
+	buffer := bytes.NewBuffer([]byte{})
+
+	StdXPipe := (*cmd).StdoutPipe
+	if pipeName == "Stderr" {
+		StdXPipe = (*cmd).StderrPipe
+	}
+
+	pr, err := StdXPipe()
+	require.NoErrorf(t, err, "Unexpected failure in call to (*Cmd).%s", pipeName)
+
+	// Goroutine will be released on (*cmd).Wait
+	go func() {
+		t.Helper()
+		n, err := io.Copy(buffer, pr)
+		assert.NoErrorf(t, err, "async error copying %s after %d bytes", pipeName, n)
+	}()
+	_, err = StdXPipe()
+	require.Errorf(t, err, "Unexpected success calling (*Cmd).%s twice", pipeName)
+
+	return buffer
+}
+
 func TestCommandOutPipes(t *testing.T) {
 	d := newTestDistro(t, rootFs)
 
+	type stream int
+	const (
+		null stream = iota
+		buffer
+		file
+	)
+
 	testCases := map[string]struct {
 		cmd    string
-		stdout bool
-		stderr bool
+		stdout stream
+		stderr stream
 
-		want string
+		wantInFile   string
+		wantInBuffer string
 	}{
-		"all discarded":           {},
-		"piped stdout":            {stdout: true, want: "Hello stdout\n"},
-		"piped stderr":            {stderr: true, want: "Hello stderr\n"},
-		"piped stdout and stderr": {stdout: true, stderr: true, want: "Hello stdout\nHello stderr\n"},
+		"all discarded": {},
+
+		// Writing to buffer
+		"stdout to a buffer":            {stdout: buffer, wantInBuffer: "Hello stdout\n"},
+		"stderr to a buffer":            {stderr: buffer, wantInBuffer: "Hello stderr\n"},
+		"stdout and stderr to a buffer": {stdout: buffer, stderr: buffer, wantInBuffer: "Hello stdout\nHello stderr\n"},
+
+		// Writing to file
+		"stdout to file":            {stdout: file, wantInFile: "Hello stdout\n"},
+		"stderr to file":            {stderr: file, wantInFile: "Hello stderr\n"},
+		"stdout and stderr to file": {stdout: file, stderr: file, wantInFile: "Hello stdout\nHello stderr\n"},
+
+		// Mixed
+		"stdout to file, stderr to buffer": {stdout: file, stderr: buffer, wantInFile: "Hello stdout\n", wantInBuffer: "Hello stderr\n"},
+		"stdout to buffer, stderr to file": {stdout: buffer, stderr: file, wantInFile: "Hello stderr\n", wantInBuffer: "Hello stdout\n"},
 	}
 
+	tmpDir := t.TempDir()
+
 	for name, tc := range testCases {
 		tc := tc
 		t.Run(name, func(t *testing.T) {
 			cmd := d.Command(context.Background(), "echo 'Hello stdout' >&1 && sleep 1 && echo 'Hello stderr' >&2")
 
-			var buff bytes.Buffer
-			if tc.stdout {
-				cmd.Stdout = &buff
+			bufferRW := &bytes.Buffer{}
+			fileRW, err := os.CreateTemp(tmpDir, "log_*.txt")
```

Comment:
We are not reusing tmpDir outside, really, forget about micro optimization in tests like this, we are creating files and so on. Let’s have a dedicated temp directory, so that the day we need to debug it, we know which file to look and not "oh, what was the log file corresponding to this test already? They are all in the same folder…")

Reply:
Fair.